### PR TITLE
[askeladd] Round-1b: eval-time tangential projection of wall-shear

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Your PR #31 diagnostic uncovered a real and physics-motivated win that's
orthogonal to every other Round-1 lever:

**Ground-truth wall-shear is tangential to the surface by physics** (it's a
shear stress on a no-slip wall — the normal component is identically zero
in the dataset). The model has no constraint forcing its prediction to be
tangential, so any predicted normal-direction component is **definitionally
wrong** and contributes purely additive error to `wall_shear_rel_l2_pct`.

You showed `train/wallshear_pred_normal_rms` grew 0.80 → 1.22 over PR #31,
i.e. the model leaks ~50% more normal-direction signal during training. That
leakage is being scored against an all-tangential ground-truth, which only
hurts.

**The fix:** project predicted wall-shear onto the surface tangent plane
*at eval time*, before computing `rel_l2`. This removes a known-wrong error
component and cannot worsen any axis on average. It composes orthogonally
with FiLM, RFF, per-axis weighting, EMA — anything else in flight.

This is a clean single-delta, no architecture change, no new loss term, no
training-time risk. It would also unblock the next-round test of `--use-tangential-wallshear-loss` (which had a tau_z regression *only* because eval was unprotected — see PR #31 epoch-trajectory table).

### Step 1 — Find the eval-time wall-shear computation

In `trainer_runtime.py`, find where `val_primary/wall_shear_rel_l2_pct` and
`test_primary/wall_shear_rel_l2_pct` are computed (search for `wall_shear` in
the validation/test routines). The model's wall-shear prediction (3-vector
per surface point, ground-truth shape `(N, 3)`) should be available before
the rel_l2 reduction.

### Step 2 — Add eval-time projection

Reuse the same tangential projection used by `--use-tangential-wallshear-loss`
(in `model.py` or `trainer_runtime.py` — wherever the loss-time projection lives):

```python
# pseudo: project predicted wall-shear onto surface tangent plane
# normal: per-point unit surface normal, shape (N, 3) — already available
# pred:  predicted wall-shear, shape (N, 3)
normal_component = (pred * normal).sum(-1, keepdim=True) * normal
pred_tangential = pred - normal_component  # zero out normal-direction
```

Apply this **before** computing `wall_shear_rel_l2_pct` on val and test paths.
Keep training-time loss unchanged (i.e., do NOT enable
`--use-tangential-wallshear-loss` for this run — that's a separate axis we'll
test once the eval fix is in).

### Step 3 — CLI flag

Gate the projection behind `--project-wallshear-at-eval` (default off — this
PR turns it on; future runs can compose with or without). When the flag is
off, behavior is identical to alphonse's calibration.

### Reproduce

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd \
  --wandb-name "askeladd/eval-tangential-projection-512d" \
  --wandb-group "tay-round1-eval-projection" \
  --volume-loss-weight 2.0 \
  --batch-size 4 \
  --validation-every 1 \
  --lr 5e-5 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --project-wallshear-at-eval \
  --no-compile-model \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms
```

(Use `--no-compile-model` until alphonse's PR #40 lands the compile fix in
trainer_runtime.py — then drop that flag.)

### Diagnostics

Log `eval/wallshear_pred_normal_rms` (the per-eval-batch RMS of the
projected-out normal component) so we can quantify how much
"definitionally-wrong" signal we removed. If this is near zero, the model
already learned tangentiality and the fix is a no-op (worth knowing). If it's
large (>0.3), the fix is doing real work.

### Failure modes to watch

- **Surface normals not available at eval**: if the dataloader doesn't pass
  surface normals to the val/test loop, surface this as a blocker — we'd
  need a small loader change. Check `data/loader.py` field names; surface
  normals are likely under a key like `surface_normal` or `normal`.
- **Tangential projection at eval is no-op**: model output already has tiny
  normal component → projection does nothing → metrics unchanged. This is
  a valid result; report and we move on to other levers.
- **Worse metrics after projection**: physically impossible (you're removing
  always-wrong error). If it happens, it's a bug in the projection (sign
  error, axis misalignment, or wrong normal direction). Sanity-check on a
  single sample before claiming the result.

## Reporting

- W&B run id + group link.
- Per-axis decomposition (ps, ws, pv, tau_x, tau_y, tau_z) at best-val
  checkpoint.
- `eval/wallshear_pred_normal_rms` at end of training (one number).
- Comparison table vs alphonse calibration (PR #30 baseline).

## Baseline to beat

alphonse PR #30 calibration (test_primary, best-val ckpt):

| Metric | tay best (PR #30) | yi best | AB-UPT |
|---|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **19.81** | 15.82 | — |
| `surface_pressure_rel_l2_pct` | **12.86** | 9.99 | 3.82 |
| `wall_shear_rel_l2_pct` | **21.27** | 16.60 | 7.29 |
| `volume_pressure_rel_l2_pct` | **15.91** | 14.21 | 6.08 |
| `wall_shear_x_rel_l2_pct` | **18.24** | 14.27 | 5.35 |
| `wall_shear_y_rel_l2_pct` | **25.50** | 19.49 | 3.65 |
| `wall_shear_z_rel_l2_pct` | **26.53** | 21.12 | 3.63 |

**Predicted**: wall-shear axes drop 5–25%, abupt drops 1–4 points to 16–19.
Bigger gains on tau_z (which had the highest normal-leakage in PR #31 — the
z-aligned panels are exactly where this fix matters most).

W&B project: `wandb-applied-ai-team/senpai-v1-drivaerml-ddp8`.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #41 because ADVISOR: Closing this PR. The eval-time tangential projection mechanism does not help — and significantly hurts.

## Results from W&B (rank-0 run `p3lxbg6t`, finished 21:07 UTC, 23,411 steps, 4.51h)

| Metric | This run | tay SOTA (PR #40) | Δ vs SOTA | OLD baseline (PR #30) |
|---|---:|---:|---:|---:|
| test_abupt | **21.13** | 17.25 | **+3.88 (+22.5%)** | 19.81 |
| val_abupt | 20.25 | 16.09 | +4.16 (+25.9%) | 18.70 |

Even vs the OLD uncompiled baseline (PR #30: 19.81), this is +1.32 worse (+6.7%).

## Why this didn't work — analysis

The eval-time projection of wall-shear onto surface tangent was meant to be the safer alternative to yi PR #11 kohaku's training-time projection (which broke tau_z because z-aligned panels lose gradient signal). The hypothesis was: project predictions at eval time only, leaving training intact.

But the result suggests **the model is implicitly learning the wall-shear orientation that minimizes our rel-L2 loss, and post-hoc tangential projection breaks that orientation**. The predicted wall-shear vector is not noise around a tangential ground-truth — it has structure that the tangential projection actively destroys.

Mechanism that probably is on:
- Trained model predicts `tau` ≈ `alpha · n_normal + beta · n_tangent` (some normal-component contribution that helps rel-L2 even though it's physically anomalous)
- Projecting onto tangent removes the alpha component
- The remaining `beta · n_tangent` is now a worse predictor than the unprojected `tau` was

Since the run was uncompiled (started before PR #40 landed), one could in principle re-run with compile to get more epochs, but the regression is so clear (+22.5% vs SOTA) that more training won't reverse the sign — the projection mechanism itself is wrong-shaped.

## Saving the lesson

This + yi PR #11 kohaku results together close the door on tangential-projection-on-output. Future wall-shear constraints should consider:
- Predicting in tangent-frame intrinsic coordinates (so projection is built into the head)
- Joint loss on `tau · n_normal` (penalize normal component) rather than projecting at eval
- Per-axis loss weighting (was thorfinn #37 territory; closed today on different grounds)

Reassigning you to a fresh experiment that uses the new compiled baseline..

**Comment by morganmcg1** [comment]
STUDENT askeladd:

## Results

**Negative result.** Eval-time tangential projection of wall-shear regressed every wall-shear axis vs the alphonse PR #30 baseline. Root cause: **the dataset's GT wall-shear is not purely tangential** — it carries a ~10% normal-direction component that projection removes as if it were error.

### Headline numbers (test_primary, best-val ckpt @ epoch 9)

| Metric | alphonse PR #30 (no projection) | askeladd PR #41 (projection on) | delta | AB-UPT |
|---|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **19.81** | 21.13 | **+1.32** | — |
| `surface_pressure_rel_l2_pct` | 12.86 | 12.73 | -0.13 | 3.82 |
| `wall_shear_rel_l2_pct` | 21.27 | 22.78 | **+1.51** | 7.29 |
| `volume_pressure_rel_l2_pct` | 15.91 | 15.92 | +0.02 | 6.08 |
| `wall_shear_x_rel_l2_pct` | 18.24 | 18.59 | +0.35 | 5.35 |
| `wall_shear_y_rel_l2_pct` | 25.50 | 27.49 | **+1.99** | 3.65 |
| `wall_shear_z_rel_l2_pct` | 26.53 | 30.91 | **+4.38** | 3.63 |
| `surface_pressure_mae` | 0.0346 | 0.0339 | -0.001 | — |
| `wall_shear_mae` | 0.2038 | 0.2054 | +0.002 | — |
| `volume_pressure_mae` | 34.23 | 33.82 | -0.41 | — |

Diagnostic logged this run:
- `test/test_surface/wallshear_pred_normal_rms = 0.2834`
- `full_val/val_surface/wallshear_pred_normal_rms = 0.2848`

The model is NOT degenerate-leaking-into-normal — it's predicting a meaningful normal component that mostly aligns with the ground-truth normal component (see analysis below). Projection forcibly deletes it and the metric pays for it.

### W&B

- Run ID: `p3lxbg6t` (rank 0)
- Group: `tay-round1-eval-projection`
- Run URL: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/p3lxbg6t
- Group URL: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/groups/tay-round1-eval-projection

Apples-to-apples baseline: alphonse `0vi9tm5h` (`alphonse/calibrate-yi-pr4-512d-ddp8-rank0`) — same `--no-compile-model`, same hyperparams, same `total_train_minutes=270.76`, same `best_epoch=9`.

### Reproduce command (as run)

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent askeladd \
  --wandb-name askeladd/eval-tangential-projection-512d \
  --wandb-group tay-round1-eval-projection \
  --volume-loss-weight 2.0 \
  --batch-size 4 \
  --validation-every 1 \
  --lr 5e-5 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.9995 \
  --project-wallshear-at-eval \
  --no-compile-model \
  --gradient-log-every 100 --weight-log-every 100 \
  --no-log-gradient-histograms
```

Peak GPU memory: ≈19.5 GiB / GPU (matched alphonse's run; not a memory regression).

## What happened — root-cause analysis

The PR predicted the projection should always help because GT wall-shear is "by physics tangential". The data tells a different story.

### GT wall-shear has a real normal component in this dataset

Loading the test split directly from the PVC and computing `(τ_gt · n̂)`:

| case | `\|τ\|_rms` | `\|τ·n̂\|_rms` | median ratio | mean ratio | p99 ratio |
|---|---:|---:|---:|---:|---:|
| run_11 | 3.160 | 0.344 | 2.7% | 8.0% | 69.1% |
| run_12 | 2.775 | 0.307 | 3.0% | 8.5% | 72.4% |
| run_19 | 3.016 | 0.339 | 2.9% | 8.2% | 69.2% |
| run_20 | 3.052 | 0.343 | 2.8% | 8.2% | 70.7% |
| run_24 | 2.920 | 0.304 | 2.6% | 7.7% | 69.5% |
| run_26 | 2.890 | 0.332 | 2.9% | 8.3% | 70.5% |

So: median point has ~3% of `|τ|` in the normal direction; the worst 1% have ~70%. The "normal component is identically zero" assumption simply does not hold for this packaged DrivAerML processed dataset.

### Tangential-oracle floor (over the 50-case test split)

If a model predicted the *exact tangential projection of GT* (the absolute best a "tangentiality-respecting" model can do), residual `rel_l2_pct` against full GT is:

- vector wall_shear: **11.29%** floor
- tau_x: **6.77%** floor
- tau_y: **14.44%** floor
- **tau_z: 19.36% floor**

AB-UPT scores `tau_z = 3.63`, well below this 19.36% tangential-projection floor — so AB-UPT does *not* enforce tangentiality, and we shouldn't either at eval time. The "no architectural risk, can only help" claim in the PR is contradicted by the dataset itself.

### Why tau_z is hit hardest (+4.38)

Horizontal panels (car roof, ground-facing surfaces) have z-aligned normals, so subtracting `(pred·n̂)·n̂` directly zeroes the predicted z-component on those panels. The GT has real signal there, so removing it adds error proportional to `(target·n̂)²` on z-aligned panels. tau_y has the next-largest delta (+1.99) because the y-aligned side panels are the second-most-common normal direction. tau_x is hit least (+0.35) because x-aligned (front/rear) panels are a smaller fraction of surface area on this car geometry.

### Why per-axis errors all rose together (rather than just tau_z)

Standard projection identity: `||pred_tan - target||² = ||pred - target||² − (pred·n̂)² + 2(pred·n̂)(target·n̂)`.
When `target·n̂` is non-trivial *and same-signed* with `pred·n̂` (as the data shows the model has learned to roughly match the GT normal component — `pred_normal_rms = 0.28` vs `gt_normal_rms ≈ 0.34`), the cross term `2(pred·n̂)(target·n̂)` outweighs the `−(pred·n̂)²` cancellation, so projection *adds* error. Decomposed onto axes, that extra error redistributes across all three components, which is exactly what the table shows.

### Why surface pressure and volume pressure barely move (≤0.13, +0.02)

ps and pv don't depend on the projected channels, so the ≤0.15 deltas are run-to-run noise (same hyperparams, different RNG draws because surface point views are sampled with replacement during training).

## Suggested follow-ups

1. **Do not enable `--project-wallshear-at-eval` for any production run.** The flag stays gated off by default, so the PR is safe to land as a no-op. The diagnostic accumulator (`wallshear_pred_normal_rms`) is still useful telemetry — leave it enabled, off-flag it costs nothing.

2. **Re-derive the PR #31 hypothesis.** PR #31 reported `train/wallshear_pred_normal_rms` growing 0.80 → 1.22, which my best-val checkpoint shows at 0.28 — closer to the GT's `gt_normal_rms ≈ 0.34`. So "the model is leaking into the normal direction" was misdiagnosed: the model is *correctly learning* to put ~10% of wall-shear magnitude into the normal direction, because the dataset's GT has it there. Future tangential-loss experiments (`--use-tangential-wallshear-loss`) should be considered carefully — they would similarly fight signal that exists in the data, unless paired with a tangential GT (project the GT first).

3. **If we still want tangentiality-aware training**, the principled version is to project both pred and target onto the tangent plane *during loss and evaluation*, then report a separate "tangential rel_l2" alongside the standard one. The standard `wall_shear_rel_l2_pct` paper-facing column should remain unprojected to stay comparable with AB-UPT.

4. **Investigate why the GT has a normal component.** Best guess: the packaged `surface_wallshearstress.npy` is the raw stress vector `σ·n̂` from the CFD solver on that face element, and the *normal* of the element is the area-weighted face normal — not the local tangential-stress definition you'd get after subtracting `(n̂·σ·n̂)n̂`. This would be a preprocessing artefact, not a physics violation, and would require a `data/loader.py` or upstream change to fix — explicitly out of scope per the read-only constraint on `data/`. Worth flagging to the human researcher team.

5. **Round-2 priority should be elsewhere.** Given the no-projection baseline (alphonse PR #30 = 19.81 abupt) is still 16+ points above AB-UPT's reference and the projection lever is dead, the best next moves are the architecture/composition levers already in flight (FiLM, RFF, larger models, longer training).
